### PR TITLE
UL2 with GPT-style Megatron models

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds the following:

1. `MegatronUGPTModel` which subclasses `MegatronGPTModel` and overrides dataset building, attention mask changes etc.
2. `UGPTDataset` which subclasses UL2Dataset and implements GPT-specific data processing.
3. Megatron U-GPT pre-training/fine-tuning script and YAML config.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ inference:
   top_k: 0  # The number of highest probability vocabulary tokens to keep for top-k-filtering.
   top_p: 0.9 # If set to float < 1, only the most probable tokens with probabilities that add up to top_p or higher are kept for generation.
   temperature: 1.0 # sampling temperature
-  add_BOS: True # add the bos token at the begining of the prompt
+  add_BOS: False # add the bos token at the begining of the prompt
```
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:6]: do we need to change this?

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml [hunk 2]
```diff
@@ -26,12 +26,13 @@ checkpoint_dir: null # checkpoint file dir. This is used to load the PTL checkpo
 checkpoint_name: null # PTL checkpoint file name, only used for PTL checkpoint loading
 hparams_file: null # model configuration file, only used for PTL checkpoint loading
 prompts: # prompts for GPT inference
-  - "Q: How are you?"
-  - "Q: How big is the universe?"
+  - "<extra_id_x>Q: How is <extra_id_0> created? A:"
+  - "<extra_id_x>Q: How big <extra_id_0> at night? A:"
```
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:30]: It seems painful to add `<extra_id_x>`? I think most people will want to just add their prompt in language?

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml (hunk 1 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:6]: Yes, for example all our recent models will do really poorly with `add_BOS=True` because the model never sees BOS. This just happens to be fine for the GPT2 tokenizer because bos = eos = |<endoftext>|.

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml (hunk 2 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:30]: Yup! I think we need to revert this before actually merging.
